### ISSUE TITLE
A few tidbits

### ISSUE THREAD

**Issue by Spirit532** [pcnorden/ReSteel#1]
I'm curious about this too, so I did a bit of work.
[Here's a gist](https://gist.github.com/Spirit532/3390d754f450b090783d074c1e7bab56) that shows how the drawing protocol works.
It lets you use the lower 128x40px of the screen to display whatever you want.

Regarding battery status: it appears to be readable by doing this:
```c
unsigned char req[31] = { 0x40, 0xAA };
res = hid_write(device_handle, req, 31);
if (res < 0)
return -1;

unsigned char data_read[2];
res = hid_read(device_handle, data_read, 1);
if (res < 0)
return -1;

int bat_status = data_read[0];
```

Where bat_status is a percentage in 0x00 to 0x04.
However, it seems to not work fully standalone, as I can only get readings once I let the SteelSeries GG software(win only) do something over USB. The readings are also wrong most of the time, which is confusing.

No idea how to detect battery charger status yet.

**Comment by pcnorden** [comment]
This is just absolutely excellent! Thank you very much for contributing this!

I'm experimenting with getting the arctis pro wireless base station to actually take commands since I've also hit the roadblock of SteelSeries GG engine needing to talk some first with the base station before I can start sending some of my own data.

I'm gonna get out my logic analyzer tomorrow and physically sniff the USB traffic to see if I can get any more concrete USB data and see what happens when I replay the data. This is due to windows trying to be too helpful sometimes with the USB stack and just rewriting data sometimes

**Comment by Spirit532** [comment]
You may find this useful, it's how I got the video sending to work.
https://github.com/JohnDMcMaster/usbrply

**Comment by pcnorden** [comment]
Honestly, thank you so much for finding that project! I will definitely use that project to start testing out how to initialize the communication on the base station!

Also, would it be alright if I rewrote the code you made for windows to draw on the screen into code that would work on linux and commit to my repository here since I would find that really useful to test out if I've got the base station into communication mode?*

I'm also really sorry for being to late to respond, but work and now vacation have and will mess with my available spare time, but I should hopefully be able to spend more time soon on this project.
Just so you know, I am also extremely thankful that you are here dropping such valuable information on me with links to code and projects that will help me extremely much on this :heart:

_* By rewriting I mean that I will implement crude error handling and write a udev rule so users/developers don't need to run the software as sudo since that just seems hacky_

**Comment by Spirit532** [comment]
Sure, feel free to do so. But keep it crossplatform :)
Side note: the drawing code _always_ works, even if the station is freshly connected with no GG. It's the other info that's inaccessible.

**Comment by pcnorden** [comment]
Sorry for the long radio silence, have been out in a tent without any good reception to update this issue.

I'll try to keep my solution cross-platform, but compiling code on my windows VM usually ends with a bluescreen or just freezing so I won't test builds too much on windows until I really start to understand the protocol for the base station a bit more than I understand it right now.

Side note: for me I actually need to replay the USB traffic for the drawing code to start working, but I believe this might be because I'm using a really old base station firmware or something, but I'm not 100% sure what is causing it since I only had a few hours to test stuff before I went out camping.

**Comment by pcnorden** [comment]
Sorry to add this to the issue, but what firmware version is your base station running?

I've made some progress in reverse-engineering what is going on, but the drawing code is just a absolute mess for some reason since it just appears to be really really buggy, so I'm guessing that I have a old firmware version or something on my base station.

**Comment by Spirit532** [comment]
I've got MCU 1.17.00 and TX 1.13.00, latest, updated through the Engine.

**Comment by nickarcade** [comment]
Is there still interest for this project? I'm interested in finding people in reverse engineering the audio devices such as the GameDAC and GameDAC gen 2. I have been performing device teardowns and have been posting the boards online since they are not found anywhere else. I have done the GameDAC, gen 2, and the Arctis Wireless base station. First two devices are STM32 devices. Happy to send you info upon your request.

**Comment by Spirit532** [comment]
Unfortunately I put my efforts on pause right around the time of my last post, and my base station OLED is so burned in that there's not much point in doing anything now. It will probably come down to just reverse-engineering the GG-specific protocol bits that I missed to make it work.

**Comment by nickarcade** [comment]
Is this project something along the lines of what you're looking for? https://github.com/Mark7888/steelseries-sonar-py

**Comment by Spirit532** [comment]
No, that uses their own API. This project is to bypass it completely and reverse-engineer the low level USB protocol.
